### ISSUE TITLE
getMaterialPropertyOld marks the property as regular

### ISSUE THREAD

**Issue by hugary1995** [idaholab/moose#19116]
## Bug Description

I have an object A that retrieves an old material property in its initialization list
```cpp
_strain_old(getMaterialPropertyOldByName<RankTwoTensor>("total_strain"))
```
and an object B that declares an AD material property as
```cpp
_total_strain(declareADProperty<RankTwoTensor>("total_strain"))
```
Then I get an error message like this
```
*** ERROR ***
Attempting to declare AD material property total_strain, but it is already retrieved/declared as a regular material property.
```

## Steps to Reproduce

A MWE [here](https://github.com/hugary1995/raccoon/blob/62c9841c3c73ffa79afa9a5dac0c5f508fc7aa24/tutorials/hydraulic_fracturing_SPF/test.i).

## Impact

This bug leads to errors that are construction order dependent.

**Comment by dschwen** [comment]
Ok, @simopier just ran into this issue again. We might have to update that API to inlcude an `is_ad` template parameter. Or did you find a workaround?

**Comment by dschwen** [comment]
No, the requestor cannot possibly know whether the current state of the property needs to be AD or non-AD. Fetching an old state just cannot trigger the creation of a material property. This is going to be tough to resolve.

**Comment by simopier** [comment]
> No, the requestor cannot possibly know whether the current state of the property needs to be AD or non-AD. Fetching an old state just cannot trigger the creation of a material property. This is going to be tough to resolve.

It might not be appropriate all the time, but could we assume that the current state of material property is the same type (AD or non-AD) as the material properties that the requestor would create?

Meaning, using the case above, object A would assume that object B would have the same type as object A. 
This is basically what it currently does anyway, but only for the non-AD case.

**Comment by hugary1995** [comment]
> Or did you find a workaround?

I didn't find any workaround.

**Comment by hugary1995** [comment]
> No, the requestor cannot possibly know whether the current state of the property needs to be AD or non-AD. Fetching an old state just cannot trigger the creation of a material property. This is going to be tough to resolve.

One design is to queue all the stateful requestors and wait to process all of them at the very of the construction phase. That's the easiest solution I can think of at the moment.

**Comment by YaqiWang** [comment]
What is this object A, a material? When using actions to add the object, developers will have control on when to add the user object but not always by MOOSE input syntax. In Griffin, we have a new task like `add_aux_materials`, which happens after `add_materials` and actions add some auxiliary materials after materials added through MOOSE syntax.

**Comment by dschwen** [comment]
OK, after a quick discussion with @loganharbour we came up with a way forward that I'll implement.

**Comment by lindsayad** [comment]
I look forward to seeing it
